NiFi-5102 - Adding Processors for MarkLogic DB

vivekmuniyandi opened the pull request:
…tMarkLogic

Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?
Yes, there is a JIRA ticket - https://issues.apache.org/jira/browse/NIFI-5102

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
Yes

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?
Yes

- [x] Is your initial contribution a single, squashed commit?
No

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
Yes, but certain tests from other processors fail - such as from grpc processors were failing with an unapproved license error for auto-generated files in the target folder.

- [x] Have you written or updated unit tests to verify your changes?
Yes

- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
Yes

- [x] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
No
- [x] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
Yes

Have added a NOTICE file and a LICENSE file under the nifi-marklogic-bundle project

- [x] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?
Yes

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?
Yes

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

MikeThomsen commented on nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/com/marklogic/nifi/processor/AbstractMarkLogicProcessor.java:
```diff
@@ -0,0 +1,104 @@
+    protected Set<Relationship> relationships;
+
+    // NiFi requires a validator for every property, even those that don't need any validation
+    protected static Validator NO_VALIDATION_VALIDATOR = new Validator() {
```
`Validator.VALID` does the same thing.

MikeThomsen commented on nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/com/marklogic/nifi/processor/AbstractMarkLogicProcessor.java:
```diff
@@ -0,0 +1,104 @@
+        .name("Thread count")
+        .displayName("Thread count")
+        .required(true)
+        .defaultValue("16")
```
NiFi uses a lot of threads, so you might want to think about lowering this default so it doesn't risk getting greedy.

MikeThomsen commented on nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/com/marklogic/nifi/processor/QueryMarkLogic.java:
```diff
@@ -0,0 +1,145 @@
+@Tags({"MarkLogic"})
+@CapabilityDescription("Creates FlowFiles from batches of documents, matching the given criteria," +
+    " retrieved from a MarkLogic server using the MarkLogic Data Movement SDK (DMSDK)")
+public class QueryMarkLogic extends AbstractMarkLogicProcessor {
```
You might want to think about setting `@InputRequirement` to ALLOWED so that you can add an incoming relationship for having the queries provided by flowfile properties or content. There are examples of how to do this in `GetMongo` and `ExecuteSQL`. When looking there, take note of the top of their `onTrigger` methods to see how the presence of a connection is detected.

vivekmuniyandi replied:
We have added this as a task in our sprint and accommodate in the coming weeks.

MikeThomsen commented on nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/com/marklogic/nifi/processor/QueryMarkLogic.java:
```diff
@@ -0,0 +1,145 @@
+        job.setWhereUriPattern(context.getProperty(URI_PATTERN).getValue());
+        job.setWhereUrisQuery(context.getProperty(URIS_QUERY).getValue());
+
+        // TODO Replace ExportListener with custom listener so that we can do a commit per batch
```
I would strongly recommend doing this TODO. `GetMongo` follows a similar pattern and it really saves the user time and resources to have a batch of documents written instead of 1:1 with a flowfile. For the loads MarkLogic is known to handle, that could be a significant volume of flowfiles, especially if a user uses this to do the equivalent of a full table fetch.

MikeThomsen commented on nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/com/marklogic/nifi/processor/QueryMarkLogic.java:
```diff
@@ -0,0 +1,145 @@
+            } catch (final Throwable t) {
+                getLogger().error("{} failed to process due to {}; rolling back session", new Object[]{this, t});
+                session.rollback(true);
+                throw t;
```
You should throw a `ProcessException` here instead.

MikeThomsen commented on nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/com/marklogic/nifi/processor/PutMarkLogic.java:
```diff
@@ -0,0 +1,382 @@
+ * The TriggerWhenEmpty annotation is used so that this processor has a chance to flush the WriteBatcher when no
+ * flowfiles are ready to be received.
+ */
+@Tags({"MarkLogic"})
```
Should add a few more descriptive tags to help people find it.

MikeThomsen commented on nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/com/marklogic/nifi/processor/PutMarkLogic.java:
```diff
@@ -0,0 +1,382 @@
+        .name("Job Name")
+        .displayName("Job Name")
+        .description("Name for the WriteBatcher job")
+        .addValidator(NO_VALIDATION_VALIDATOR)
```
Change to Validator.VALID

MikeThomsen commented on nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/com/marklogic/nifi/processor/PutMarkLogic.java:
```diff
@@ -0,0 +1,382 @@
+        .name("Temporal collection")
+        .displayName("Temporal collection")
+        .description("The temporal collection to use for a temporal document insert")
+        .addValidator(NO_VALIDATION_VALIDATOR)
```
This should use StandardValidators.NON_BLANK_VALIDATOR if you need it to be present. If not, use Validator.VALID.

MikeThomsen commented on nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/com/marklogic/nifi/processor/PutMarkLogic.java:
```diff
@@ -0,0 +1,382 @@
+        }
+    }
+    private Map<String, FlowFileInfo> URIFlowFileMap = new HashMap<>();
+    public static final PropertyDescriptor COLLECTIONS = new PropertyDescriptor.Builder()
```
Rule of thumb for all of these:

1. Add an explicit call to `required(boolean)` so it's obvious what your intent is there.
2. Switch `NO_VALIDATION_VALIDATOR` to Validator.VALID.

MikeThomsen commented on nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/com/marklogic/nifi/processor/PutMarkLogic.java:
```diff
@@ -0,0 +1,382 @@
+    public static final PropertyDescriptor URI_PREFIX = new PropertyDescriptor.Builder()
+        .name("URI prefix")
+        .displayName("URI prefix")
+        .description("(Optional) The prefix to prepend to each URI")
```
Remove `(Optional)` and replace with an explicit `required(boolean)` declaration.

MikeThomsen commented on nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/com/marklogic/nifi/processor/PutMarkLogic.java:
```diff
@@ -0,0 +1,382 @@
+    public static final PropertyDescriptor URI_SUFFIX = new PropertyDescriptor.Builder()
+        .name("URI suffix")
+        .displayName("URI suffix")
+        .description("(Optional) The suffix to append to each URI")
```
Same here.

MikeThomsen commented on nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/com/marklogic/nifi/processor/PutMarkLogic.java:
```diff
@@ -0,0 +1,382 @@
+        .addValidator(NO_VALIDATION_VALIDATOR)
+        .build();
+
+    protected static final Relationship SUCCESS = new Relationship.Builder()
```
Nit: both of these descriptions should have a `.` at the end of them.

MikeThomsen commented on nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/com/marklogic/nifi/processor/PutMarkLogic.java:
```diff
@@ -0,0 +1,382 @@
+            this.session = session;
+        }
+    }
+    private Map<String, FlowFileInfo> URIFlowFileMap = new HashMap<>();
```
This variable name violates standard Java conventions.

MikeThomsen commented on nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/com/marklogic/nifi/processor/PutMarkLogic.java:
```diff
@@ -0,0 +1,382 @@
+        DocumentMetadataHandle metadata = (DocumentMetadataHandle) writeEvent.getMetadata();
+        String flowFileUUID = metadata.getMetadataValues().get("flowFileUUID");
+        FlowFileInfo flowFile = URIFlowFileMap.get(flowFileUUID);
+        if(flowFile != null) {
```
Have you tried this with a very large data set to see how it performs? Like millions of sample XML records?

MikeThomsen commented:
@vivekmuniyandi looks like you pulled a few other folks' commits in with your last push. Do this to clear that up:

1. git checkout master
2. git pull <apache remote here> master
3. git checkout nifi-5102
4. git rebase master
5. git push marklogic --force nifi-5102

You probably did a pull on master into nifi-5102. You want to avoid that for your own sanity's sake and use a rebase instead.

vivekmuniyandi commented:
I followed this link - https://cwiki.apache.org/confluence/display/NIFI/Contributor+Guide#ContributorGuide-Keepingyourfeaturebranchcurrent

and looks like this pulled other's commits as well.

vivekmuniyandi commented:
Thanks @MikeThomsen ! Have made the changes.

MikeThomsen commented:
@vivekmuniyandi I'll try to build a MarkLogic Docker image and share it on Docker Hub so others can use that if they want.

MikeThomsen commented:
BTW, I would strongly recommend your team discuss adding a `PutMarkLogicRecord` processor so you can do a bulk ingestion invent from a single flowfile. We have quite a few good implementations such as ones for HBase, ElasticSearch and MongoDB that you can use/steal from to make it happen. Would **strongly** recommend you do that because it'll make bulk ingestion of very large data sets go much faster for MarkLogic. If you want to do that, feel free to just start work on it and push changes into this PR and we'll just keep going.

MikeThomsen commented:
@vivekmuniyandi I'll try to find time to set up a MarkLogic node for testing some time in the next few days (day job and such is getting in the way). In the mean time, I would suggest reaching out to @joewitt directly to see if he or any of his folks have any time they can spare to jump in and help you out. Also, just a suggestion, you might want to think about setting up a secure cluster that you can privately allow reviewers to access so we/they can work with you to confirm everything works the way you expect.

vivekmuniyandi commented:
@MikeThomsen We don't have a public MarkLogic Docker image but we do have [this](https://hub.docker.com/r/patrickmcelwee/marklogic-dependencies/) on Docker Hub which would give you a head start on having a MarkLogic instance up and running. 

I will drop them an email and I will work on getting a secure access to the cluster. Thanks for all the help.

joewitt commented on nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-nar/pom.xml:
```diff
@@ -0,0 +1,56 @@
+        <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-marklogic-services-api-nar</artifactId>
+            <version>${project.parent.version}</version>
```
i strongly recommend avoiding use of ${} notation for version of components being built within this reactor/multi-module build of nifi.  It can create problems during the release process and our release plugins take care of changing the strings for us.  Please use '1.7.0-SNAPSHOT' instead.

joewitt commented on nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/pom.xml:
```diff
@@ -0,0 +1,88 @@
+        <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-utils</artifactId>
+            <version>${nifi.version}</version>
```
i strongly recommend avoiding use of ${} notation for version of components being built within this reactor/multi-module build of nifi.  It can create problems during the release process and our release plugins take care of changing the strings for us.  Please use '1.7.0-SNAPSHOT' instead.

joewitt commented on nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/com/marklogic/nifi/processor/PutMarkLogic.java:
```diff
@@ -0,0 +1,396 @@
+    private WriteBatcher writeBatcher;
+    // If no FlowFile exists when this processor is triggered, this variable determines whether or not a call is made to
+    // flush the WriteBatcher
+    private boolean shouldFlushIfEmpty = true;
```
It appears this member variable could benefit from being 'volatile'.  This is true for the writeBatcher and possibly dataMovementManager as well.

joewitt commented:
Ok i've attached a patch which helps with some aspects of POM construction, flagging things like resource utilization since it appears to be loading full content into memory, and renaming the service to indicate it is a MarkLogic service rather than just a database service.  There is an outstanding need to sort out the security configuration.  For SSLContext stuff those things should utilize the standard mechanism of obtaining that as you can follow from a number of other processors.  Also, there is a kerberos context for security setting but there does not appear to be any associated settings for the user.  The security configurations should be removed in favor of simple/digest for now OR completed and with some consistency to other items.  For security relevant things CVEs become a concern so we take these more seriously.  For things about the performance/logic of the processor interaction with MarkLogic that we can improve over time if needed but security we want to get right up front.  The other thing that needs to happen is the nar bundles need their LICENSE/NOTICE file(s) added if necessary.  I looked at one of the nars and there would definitely need to be entries.  Please try adding these in like other nars and I'm happy to help tweak it to get it to the finish line.

If you have questions on how to achieve any of the above please ask.  Show an example nar you looked at which is similar so that we can best help close remaining gaps but from a place of good examples that you've looked at.

Thanks

vivekmuniyandi commented:
Thanks @joewitt for the comment. I am addressing all the changes you have mentioned. I will address the SSLContext and remove the Kerberos and Certificate auth for now. 

``` The other thing that needs to happen is the nar bundles need their LICENSE/NOTICE file(s) added if necessary. I looked at one of the nars and there would definitely need to be entries.```

Wrt this, I have a [LICENSE](https://github.com/apache/nifi/pull/2671/files#diff-53deed39bf31085fbecf77ea6a2382dc) and [NOTICE](https://github.com/apache/nifi/pull/2671/files#diff-a2f6b487a7a70d5f43fa320730b2c87a) file prepared by our legal team in the root directory (to account for the contents of the root directory and the sub directories) of the nifi marklogic bundle. That constitutes for all the dependecies added in our MarkLogic bundle. Should we do something more? Can you explain a bit more here as to what is required? 

Thanks for all the help?

MikeThomsen commented:
@vivekmuniyandi Unrelated, but I don't have your email: consider adding a MarkLogicLookupService in a future sprint. You can look at HBaseLookupService and MongoDBLookupService as examples. Might be highly useful to users to be able to enrich a record set using MarkLogic. I have PRs open for some LookupService-related tasks that add some additional schema-related capabilities and those might be useful to your team on that issue.

vivekmuniyandi commented:
@joewitt I have addressed all your comments except for the License and Notice comments. Can you please let us know what more should we add apart from the LICENSE and NOTICE file prepared by our legal team which we have included in the root directory of the nar? That constitutes for all the dependecies added. What more should be added? Please help. Thanks.

vivekmuniyandi commented:
@MikeThomsen Sure, will add that to our backlog. Thanks!

joewitt commented on nifi-nar-bundles/nifi-marklogic-bundle/LICENSE.txt:
```diff
@@ -0,0 +1,72 @@
+Apache License 
```
This license file should be located in the NAR under meta-inf just like the following example.  Further, it should contain the same text/same format of text for the ALv2 license.

https://github.com/apache/nifi/blob/master/nifi-nar-bundles/nifi-azure-bundle/nifi-azure-nar/src/main/resources/META-INF/LICENSE

joewitt replied:
Also unless you're altering the license by appending items to it as would be necessary was there any copied source from somewhere and that source was from the ASF category A you dont need to include it as it will be pulled in automatically.

joewitt commented on nifi-nar-bundles/nifi-marklogic-bundle/NOTICE.txt:
```diff
@@ -0,0 +1,2106 @@
+MarkLogic NiFi Processors
+
+
+Copyright � 2018 MarkLogic Corporation.
```
This notice file should be at the same location under META-INF just like this notice file is for the azure nar

https://github.com/apache/nifi/blob/master/nifi-nar-bundles/nifi-azure-bundle/nifi-azure-nar/src/main/resources/META-INF/NOTICE

Also, it isn't Copyright MarkLogic if you're submitting this as a pull request to the Apache Software Foundation.  The notice header then should read
---
nifi-marklogic-nar
Copyright 2015-2018 The Apache Software Foundation

This product includes software developed at
The Apache Software Foundation (http://www.apache.org/).
--

Then be followed by that which is legally required by the binary dependencies of this nar bundle - no more no less.

joewitt commented on nifi-nar-bundles/nifi-marklogic-bundle/NOTICE.txt:
```diff
@@ -0,0 +1,2106 @@
+
+Copyright � 2018 MarkLogic Corporation.
+
+This project is licensed under the Apache License, Version 2.0 (the "License"); you may not use this project except in compliance with the License. You may obtain a copy of the License at
```
Please see other NOTICE files for example.  https://github.com/apache/nifi/blob/master/nifi-nar-bundles/nifi-azure-bundle/nifi-azure-nar/src/main/resources/META-INF/NOTICE

All of this ALv2 reference information does not belong here.

joewitt commented on nifi-nar-bundles/nifi-marklogic-bundle/NOTICE.txt:
```diff
@@ -0,0 +1,2106 @@
+Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+
+
+Please direct questions, comments and requests to fossreview@marklogic.com. 
```
This is not appropriate for a NOTICE within an Apache project.  Same applies to the above statement.  If these are important to Marklogic it is important to consider that you don't have to contribute your code/Nar to the apache project.  You could simply cancel the PR and JIRA and have this code available on your own company website/etc..

Finally, consider that all contributions to Apache projects are as individuals.  This is not a contribution of Marklogic the corporation or else we cannot accept it.

joewitt commented on nifi-nar-bundles/nifi-marklogic-bundle/NOTICE.txt:
```diff
@@ -0,0 +1,2106 @@
+Open source software required to be made available under license is included herein. In the event you are unable to obtain a copy of such open source software, please contact fossreview@marklogic.com and a copy will be made available to you.
+
+
+The following software may be included in this project (last updated May 1, 2018):
```
This line can be removed.  The NOTICE needs to reflect what is actually bundled when this NAR is built and should account for what is in it that required NOTICE entries no more no less.

joewitt commented on nifi-nar-bundles/nifi-marklogic-bundle/NOTICE.txt:
```diff
@@ -0,0 +1,2106 @@
+
+The following software may be included in this project (last updated May 1, 2018):
+
+Apache Commons Codec� 1.7 
```
Please have this follow formatting similar to what is shown in this example

https://github.com/apache/nifi/blob/master/nifi-nar-bundles/nifi-azure-bundle/nifi-azure-nar/src/main/resources/META-INF/NOTICE#L30

joewitt replied:
Please model all of these dependency references off examples you can find in the existing source tree.  I dont see any new dependencies here that you would have trouble finding examples of how to reference.  If you have questions on how to find any of these please let me know.  Formatting is important to help this be readable/consistent/maintainable.

joewitt commented on nifi-nar-bundles/nifi-marklogic-bundle/NOTICE.txt:
```diff
@@ -0,0 +1,2106 @@
+Source materials are available for download at: https://github.com/apache/commons-lang 
+
+
+Apache Commons Logging� 1.1.1 
```
i dont see commons logging in your nars.  Where do you see it?

joewitt commented on nifi-nar-bundles/nifi-marklogic-bundle/NOTICE.txt:
```diff
@@ -0,0 +1,2106 @@
+Source materials are available for download at: https://github.com/apache/commons-logging 
+
+
+Apache Derby 10.13.1.1 
```
i dont see apache derby in your nars.  do you?

joewitt commented on nifi-nar-bundles/nifi-marklogic-bundle/NOTICE.txt:
```diff
@@ -0,0 +1,2106 @@
+The Derby build relies on jar files supplied by the Apache Lucene
+project. The following notice covers the Lucene files:
+
+Apache Lucene
```
i dont think lucene is bundled.

joewitt commented on nifi-nar-bundles/nifi-marklogic-bundle/NOTICE.txt:
```diff
@@ -0,0 +1,2106 @@
+project. The following notice covers the Lucene files:
+
+Apache Lucene
+Copyright 2013 The Apache Software Foundation
```
ok can you please dramatically reduce the size of this NOTICE following the above advice.

1) Only include reference to things actually bundled within a given nar.  No more.
2) Only include that which is necessary and consistent with other Nars.
3) Do not include reference to other nifi things.  It is part of the nifi project/source so we're good.

joewitt commented on nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/com/marklogic/nifi/processor/AbstractMarkLogicProcessor.java:
```diff
@@ -0,0 +1,91 @@
+        .displayName("Thread count")
+        .required(false)
+        .description("The number of threads - sets the thread count on the Batcher")
+        .addValidator(StandardValidators.POSITIVE_INTEGER_VALIDATOR)
```
probably makes sense to include a default value here for people to undersatnd how many threads this will have by default

joewitt commented on nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/com/marklogic/nifi/processor/PutMarkLogic.java:
```diff
@@ -0,0 +1,399 @@
+        // to the appropriate relationship
+        String flowFileUUID = flowFile.getAttribute(CoreAttributes.UUID.key());
+        metadata.withMetadataValue("flowFileUUID", flowFileUUID);
+        // TODO Haven't had luck with wrapping the FlowFile's inputStream with an ML InputStreamHandle, so copying everything into a byte array
```
recommend removal of all commented code/sections.  Could file a follow on JIRA for such actions if desired

MikeThomsen commented:
@vivekmuniyandi you have a merge conflict now.

joewitt commented:
team; given https://github.com/marklogic/nifi/releases could we consider closing this PR and keeping the MarkLogic artifact creation/maintenance something MarkLogic takes care of at this time? It is a perfectly fine model.  We could even create a nifi web page to point at vendor/other community managed/supported extensions possibly.

ryanjdew commented:
@joewitt I appreciate the positive feedback on the process we've put in place for releasing our NiFi bundle.  I would still like to see this bundle make its way into an Apache NiFi release. I'll be taking on the responsibility for addressing issues in this PR.  After a review of the PR history, I believe previous issues referenced have been addressed. Please let me know if there are any other changes you'd like to see made.

  Also, I've noticed that the Travis CI tests can be inconsistent at times. Is this a known issue and is there anything I can do to help the matter? 

ryanjdew commented:
Following up, are there any other concerns with this PR? If needed, I can provide credentials to a MarkLogic instance for testing.

joewitt commented:
@ryanjdew We have addressed Travis-CI related tests issues and the build has been stable now since.  It could break again as people add timing dependent tests that behave wildly different on slow environments but we'll see.

This PR, assuming the L&N and such are sorted now, is just tricky because we need a committer with time to devote to learning MarkLogic enough to setup an environment and verify function or leverage a provided instance (not a sustainable model).  It is a good example where the limits of what the community can reasonably support comes into play.

Now, this said this is probably a really cool and useful thing to offer folks and beneficial to both the NiFi userbase and MarkLogic user base.  This is why I suggested MarkLogic folks just hang onto this code in some public github repo and have it be ALv2.  They can publish their nars into maven central or wherever they do and provide instructions to it.  I'd be supportive, and I'd assume the community at large would, of having links to such extensions on the apache website.  This feels to me like the best tradeoff right now for all parties.

joewitt commented:
as a good example where we should have though it through more as well we have some cool work done by the InfluxDB folks.  They're now wanting to improve it in https://github.com/apache/nifi/pull/2743

But the reality is we just don't have people knowledgeable enough to do reliable code reviews/testing of this.

ryanjdew commented:
@joewitt I understand the concern. We would like to simplify the user experience of using NiFi with MarkLogic processors. We have a current model for creating and releasing NARs via a GitHub releases page, but would you happen to have a good example of a process using Maven publishing with NARs in NiFi?

joewitt commented:
If you take a look here https://nifi.apache.org/docs/nifi-docs/html/administration-guide.html#core-properties-br and checkout the 'nifi.nar.library.directory' section you can see how someone could be guided to create a directory for custom nars, add that in there, and start up.

Once we have support for extensions in the NiFi Registry this will be beautiful/easy.

MikeThomsen commented:
For anyone who wants to use this functionality, it appears that they have set up their own bundle download [here](https://github.com/marklogic/nifi/releases). Closing.
